### ISSUE TITLE
torch.nn.parallel.DistributedDataParallel is slow on backpropagation

### ISSUE THREAD

**Issue by xzhu1900** [pytorch/pytorch#29939]
## 🐛 Bug

We are training with Bert model from [this repository](https://github.com/microsoft/AzureML-BERT/tree/master). We found out that when using `torch.nn.parallel.DistributedDataParallel`, we incur a ~3x performance slowdown per iteration compared to the default Apex DistributedDataParallel. 

We broke down the training into steps and found out the major slowdown happened at the backpropagation step. `loss.backward()` is ~10x slower for models wrapped with  `torch.nn.parallel.DistributedDataParallel` compared to `distributed_apex.DistributedDataParallel`. 
The time is measured with `torch.cuda.synchronize()` and we only use fp32 for calculation.

## To Reproduce

1. follow the [notebook ](https://github.com/microsoft/AzureML-BERT/blob/master/pretrain/PyTorch/notebooks/BERT_Pretrain.ipynb) instruction to run the model
1. Switch the DDP package from `distributed_apex.DistributedDataParallel` to `torch.nn.parallel.DistributedDataParallel` at [this line](https://github.com/microsoft/AzureML-BERT/blob/be1d2f88a8ff20792a697c92fac1ba4c42fafd57/pretrain/PyTorch/train.py#L418)
1. Run the training script based on the [config file](https://github.com/microsoft/AzureML-BERT/blob/master/pretrain/configs/bert-large.json)

## Expected behavior

With full fp32 model, `torch.nn.parallel.DistributedDataParallel` should have around the same performance compared to `distributed_apex.DistributedDataParallel`. But it is actually ~3x slower on each iteration and ~10x slower on back propagation.

## Environment
PyTorch version: 1.1.0a0+828a6a3
Is debug build: No
CUDA used to build PyTorch: 10.1.163

OS: Ubuntu 16.04.6 LTS
GCC version: (Ubuntu 5.4.0-6ubuntu1~16.04.11) 5.4.0 20160609
CMake version: version 3.5.1

Python version: 3.6
Is CUDA available: Yes
CUDA runtime version: 10.1.163
Nvidia driver version: 418.87.00
cuDNN version: /usr/lib/x86_64-linux-gnu/libcudnn.so.7.6.0

How you installed PyTorch: FROM nvcr.io/nvidia/pytorch:19.05-py3
 
## Additional context

cc @ezyang @gchanan @zou3519 @jerryzh168 @pietern @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @xush6528 @VitalyFedyunin @ngimel @mruberry

**Comment by zhaojuanmao** [comment]
@xzhu1900 thanks for reporting, we will look into it and do some analysis

**Comment by pietern** [comment]
A difference between Apex and PyTorch DDP that comes to mind here is that Apex figures out the order in which gradients are produced at runtime, whereas PyTorch DDP still assumes that gradients are produced in reverse order how they are defined in the `nn.Module`. If the model you link to defines some first encoders / MLPs / etc as the last parameters in the module, you would see head-of-line blocking where all reductions will be sequenced after the final gradients, effectively removing all opportunity for overlapping reduction with gradient computation.

**Comment by xzhu1900** [comment]
The bert model we use is from this public package: https://pypi.org/project/pytorch-pretrained-bert/

I assume the layers are defined in order, but doesn't hurt to double check.

**Comment by zhaojuanmao** [comment]
another difference btw Apex and PyTorch DDP is that Apex uses multiple CUDA streams underneath. will try to install the model and train it using Apex and PyTorch DDP in our infra and see whether it can be reproduced

**Comment by zhaojuanmao** [comment]
https://github.com/pytorch/pytorch/pull/35137 was landed to address the difference btw PT DDP and Apex DDP, it is available in nightly build now.

Please let me know how it works for your flow, thanks a lot!
